consensus/byzantine_test.go

```diff
@@ -349,6 +349,230 @@ func TestByzantineConflictingProposalsWithPartition(t *testing.T) {
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{prevote}))
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{precommitNil}))
+				}
+			}
+		} else if height == 1 && round == 1 {
+			// We sign anything our friends need!
+			prevote, _ := cs.signVote(tmproto.PrevoteType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
+			precommit, _ := cs.signVote(tmproto.PrecommitType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
+
+			for i, peer := range peers {
+				if i != VICTIM_PEER_INDEX {
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{prevote}))
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{precommit}))
+				}
+			}
+		}
+	}
+
+	css[BYZ_INDEX].decideProposal = func(height int64, round int32) {}
+	css[BYZ_INDEX].setProposal = func(proposal *types.Proposal) error {
+		if proposal.Height == 1 && proposal.Round == 0 {
+			// this is us. Let's fork the tendermint
+		} else if proposal.Height == 1 && proposal.Round == 1 {
+			// Lets help friends to make a fork
+			return css[BYZ_INDEX].defaultSetProposal(proposal)
+		} else if proposal.Height == 2 && proposal.Round == 0 {
+			// Already forked!
+		}
+		return nil
+	}
+	// ----------------------------------------------------------------------
+
+	defer func() {
+		for _, sw := range switches {
+			err := sw.Stop()
+			require.NoError(t, err)
+		}
+	}()
+
+	// start the non-byz state machines.
+	// note these must be started before the byz
+	for i := 0; i < N; i++ {
+		if i != BYZ_INDEX {
+			cr := reactors[i].(*Reactor)
+			cr.SwitchToConsensus(cr.conS.GetState(), false)
+		}
+	}
+
+	// start the byzantine state machine
+	byzR := reactors[BYZ_INDEX].(*ByzantineReactor)
+	s := byzR.reactor.conS.GetState()
+	byzR.reactor.SwitchToConsensus(s, false)
+
+	// byz proposer sends one block to peers[0]
+	// and the other block to peers[1] and peers[2].
+	// note peers and switches order don't match.
+	peers := switches[BYZ_INDEX].Peers().List()
+
+	// partition A
+	ind0 := getSwitchIndex(switches, peers[VICTIM_PEER_INDEX])
+
+	// partition B
+	ind1 := getSwitchIndex(switches, peers[0])
+	ind2 := getSwitchIndex(switches, peers[2])
+	p2p.Connect2Switches(switches, ind1, ind2)
+
+	m := 0
+	for {
+		for i := 0; i < N; i++ {
+			v := css[BYZ_INDEX].Votes.Precommits(0).GetByIndex(int32(i))
+			if v != nil {
+				if v.BlockID.Hash != nil {
+					peers[VICTIM_PEER_INDEX].Send(VoteChannel, MustEncode(&VoteMessage{v}))
+					m |= (0x1 << i)
+				}
+			}
+		}
+		time.Sleep(10 * time.Millisecond)
+		if m == 0x6 {
+			break
+		}
+	}
+
+	// wait for someone in the big partition (B) to make a block
+	<-blocksSubs[ind2].Out()
+
+	t.Log("A block has been committed. Healing partition")
+	p2p.Connect2Switches(switches, ind0, ind1)
+	p2p.Connect2Switches(switches, ind0, ind2)
+
+	// wait till everyone makes the first new block
+	// (one of them already has)
+	wg := new(sync.WaitGroup)
+	wg.Add(2)
+	for i := 1; i < N-1; i++ {
+		go func(j int) {
+			<-blocksSubs[j].Out()
+			wg.Done()
+		}(i)
+	}
```

Comment:
Would be nice if we could collect all the block together and use the hashes to prove that there was a fork at that height

Reply:
> . A fork would be when validators have committed blocks with different blockID's (at the same height).

Right. I can say you did a good job in Tendermint. Well done. Tendermint is safe and secure.
It looks I made a mistake by keeping the wrong index for pre-commit votes.

> if a node prevotes a block but then receives 2f + 1 PREVOTE messages for nil then it should pre-commit nil 

I think in this case, it's better to be silent and don't send any vote. Because there is a valid proposal. But it doesn't change anything at all. It just make a cleaner voteset.

Sorry if I made any trouble. 
At this point I close and archive this PR. Again good job and be safe.

In case you need to check the fork with log files, you can take a look at here.
You may check this branch for log files: https://github.com/tendermint/tendermint/compare/master...mantre007:fork-with-log

Have fun